nestjs/src/courses/team-distribution/distribute-students.service.ts

```diff
@@ -0,0 +1,188 @@
+import { Team, TeamDistributionStudent } from '@entities/index';
+import { TeamDistribution } from '@entities/teamDistribution';
+import { Injectable, InternalServerErrorException } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { shuffle } from 'lodash';
+import { DataSource, Repository } from 'typeorm';
+import { TeamDistributionStudentService } from './team-distribution-student.service';
+import { TeamService } from './team.service';
+
+@Injectable()
+export class DistributeStudentsService {
+  constructor(
+    @InjectRepository(TeamDistribution)
+    private repository: Repository<TeamDistribution>,
+    private teamDistributionStudentService: TeamDistributionStudentService,
+    private teamService: TeamService,
+    private dataSource: DataSource,
+  ) {}
+
+  public getById(id: number) {
+    return this.repository.findOneOrFail({ where: { id } });
+  }
+
+  private getTeamCapacity(teams: Team[], teamSize: number) {
+    return teams.reduce((acc, curr) => acc + teamSize - curr.students.length, 0);
+  }
+
+  private async removeTeams(smallTeams: Team[], teamDistributionId: number, courseId: number) {
+    const queryRunner = this.dataSource.createQueryRunner();
+    await queryRunner.connect();
+    await queryRunner.startTransaction();
+    const studentIds = smallTeams.flatMap(el => el.students.map(s => s.id));
+    try {
+      await queryRunner.manager.save(
+        Team,
+        smallTeams.map(t => ({ ...t, students: [] })),
+      );
+      await queryRunner.manager.remove(Team, smallTeams);
+      const teamDistributionStudents = await this.teamDistributionStudentService.getTeamDistributionStudents(
+        studentIds,
+        teamDistributionId,
+        courseId,
+      );
+      await queryRunner.manager.save(
+        TeamDistributionStudent,
+        teamDistributionStudents.map(el => ({ ...el, distributed: false })),
+      );
+
+      await queryRunner.commitTransaction();
+    } catch (error) {
+      await queryRunner.rollbackTransaction();
+      throw new InternalServerErrorException();
+    } finally {
+      await queryRunner.release();
+    }
+  }
```

Comment:
There's a lot of places in this file, where you do something like:
```ts
const queryRunner = this.dataSource.createQueryRunner();
await queryRunner.connect();
await queryRunner.startTransaction();
try {
     // do stuff
     await queryRunner.commitTransaction();
} catch {
      await queryRunner.rollbackTransaction();
      throw new InternalServerErrorException();
} finally {
      await queryRunner.release();
}
```
Perhaps it worth to extract this logic into something reusable to avoid duplication, because logic is pretty similar